www/checktokenchallenges.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+    try{
+        $authStateId = $_REQUEST['StateId'];
+        $state = SimpleSAML_Auth_State::loadState($authStateId, 'privacyidea:privacyidea:init');
+    } catch (Exception $e){
+    }
+
+    $serverconfig = $state['privacyidea:serverconfig'];
+    $pushToken = $_GET['token'];
+    $authToken = sspmod_privacyidea_Auth_utils::fetchAuthToken($serverconfig);
+
+    $result = false;
+
+    foreach (
+        sspmod_privacyidea_Auth_utils::curl(
+            array(),
+            array("authorization:" . $authToken),
+            $serverconfig,
+            "/token/challenges/" . $pushToken,
+            "GET")
+            ->result
+            ->value
+            ->challenges
+        as $i => $e
+    ) {
+        if (strtotime($e->expiration) - time() > 0 && $e->otp_valid) {
```

Comment:
Checking the challenge expiration date is a good idea! But I think you're right, I guess we'll need a server-side code change due to https://github.com/privacyidea/privacyidea/issues/1586

Reply:
If I understand correctly, the plugin currently fetches *all* challenges for a given serial and checks if there is any non-expired challenge with ``otp_valid=1``. Is it possible to modify this to only fetch challenges for the transaction ID of the challenge that was triggered by an earlier call to ``/validate/triggerchallenge``? This could be done by sending a ``transaction_id`` parameter, analogously to the ownCloud plugin: 
https://github.com/privacyidea/privacyidea-owncloud-app/blob/f850bd9994765542324487d847d73beadf8c244c/twofactor_privacyidea/lib/Provider/TwoFactorPrivacyIDEAProvider.php#L376

As you already noted, the expiration check does not work in all cases because of #1586. If we modify the plugin such that only check the status of challenges bound to the triggered transaction, I think it would be safe to remove the expiration check for now, until privacyidea/privacyidea#1838 is implemented: Attackers could then only authenticate using an expired challenge if they are able to provide the transaction ID, which should not be possible. When #1838 is implemented, we can change the plugin code accordingly.

Hm, this was a tricky line of thought. What do you think? If I wrote something confusing, just let me know. :)

Reply:
I just realized the expiration check isn't really critical at this point, because ``checktokenchallenges.php`` only serves as a polling endpoint anyway, and the actual authentication is handled in a subsequent request anyway, right?

Reply:
... aaand I just realized that the plugin apparently allows a user to supply the transaction_id :) So yeah, I think we should be a bit careful with deleting the expiration check.

Reply:
> I just realized the expiration check isn't really critical at this point, because `checktokenchallenges.php` only serves as a polling endpoint anyway, and the actual authentication is handled in a subsequent request anyway, right?

This is correct. Any change made here will have no impact on security whatsoever. At worst, if this check wrongly deems a user to be successfully authenticated, a `/validate/samlcheck` will be triggered for the user, even though they are not authenticated yet and an error will occur. Currently there is no proper error message for this (since it should never happen during normal use), the user will just get a generic wrong-username-or-password-style error page and be asked to login again.

So we do not need to concern ourselves with a malicious actor here, but this check shouldn't return the wrong result at any time during normal operation. The actual login code did not require any further changes and I *hope* it handles any edge cases correctly.

Reply:
I have patched this to not check the expiration for now, but I have to poke at this a little more to make sure it behaves as expected. I will push this tomorrow.

Reply:
Thanks for the explanation, sounds good :) In this case, I think we should have a closer look at the login code to ensure that it doesn't allow users to inject a transaction ID.

Reply:
Fixed for now, by removing the expiry-check.